crates/stackable-certs/src/ca/ca_builder.rs

```diff
@@ -0,0 +1,240 @@
+use stackable_operator::time::Duration;
+use tracing::debug;
+use x509_cert::{
+    builder::{Builder, CertificateBuilder, Profile},
+    der::{DecodePem, referenced::OwnedToRef},
+    ext::pkix::AuthorityKeyIdentifier,
+    name::Name,
+    serial_number::SerialNumber,
+    spki::SubjectPublicKeyInfoOwned,
+    time::Validity,
+};
+
+use super::CertificateAuthority;
+use crate::{
+    CertificatePair,
+    ca::{DEFAULT_CA_VALIDITY, PEM_LINE_ENDING, SDP_ROOT_CA_SUBJECT},
+    keys::CertificateKeypair,
+};
+
+/// Defines all error variants which can occur when creating a CA
+#[derive(Debug, Snafu)]
+pub enum CreateCertificateAuthorityError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("failed to parse validity"))]
+    ParseValidity { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse \"{subject}\" as subject"))]
+    ParseSubject {
+        source: x509_cert::der::Error,
+        subject: String,
+    },
+
+    #[snafu(display("failed to create signing key pair"))]
+    CreateSigningKeyPair { source: E },
+
+    #[snafu(display("failed to serialize public key as PEM"))]
+    SerializePublicKey { source: x509_cert::spki::Error },
+
+    #[snafu(display("failed to decode SPKI from PEM"))]
+    DecodeSpkiFromPem { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse AuthorityKeyIdentifier"))]
+    ParseAuthorityKeyIdentifier { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to create certificate builder"))]
+    CreateCertificateBuilder { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to add certificate extension"))]
+    AddCertificateExtension { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to build certificate"))]
+    BuildCertificate { source: x509_cert::builder::Error },
+}
+
+/// This builder builds certificate authorities of type [`CertificateAuthority`].
+///
+/// Example code to construct a CA:
+///
+/// ```no_run
+/// use stackable_certs::{
+///     keys::ecdsa,
+///     ca::{CertificateAuthority, CertificateAuthorityBuilder},
+/// };
+///
+/// let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+///     .build()
+///     .build_ca()
+///     .expect("failed to build CA");
+/// ```
+#[derive(Builder)]
+pub struct CertificateAuthorityBuilder<'a, SKP>
+where
+    SKP: CertificateKeypair,
+    <SKP::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    /// Required subject of the certificate authority, usually starts with `CN=`.
+    #[builder(default = SDP_ROOT_CA_SUBJECT)]
+    subject: &'a str,
+
+    /// Validity/lifetime of the certificate.
+    ///
+    /// If not specified the default of [`DEFAULT_CA_VALIDITY`] will be used.
+    #[builder(default = DEFAULT_CA_VALIDITY)]
+    validity: Duration,
+
+    /// Serial number of the generated certificate.
+    ///
+    /// If not specified a random serial will be generated.
+    serial_number: Option<u64>,
```

Comment:
note: This is an internal detail we shouldn't allow being customized by the user of this library. I'm sure there is something like

```suggestion
    /// Serial number of the generated certificate.
    #[builder(skip)]
    serial_number: Option<u64>,
```

Reply:
Agreed. On that topic, if we want to have proper CAs... then serial should increment.
So, if we restore a CA (eg: on startup, read from a secret to get the CA keypair), we also need information for the certs that have been issued (not by reading available signed certs, since they might be gone already). We basically need an atomic counter.

I'm not 100% certain of the downside if we just always issue the same version though.

Reply:
e9e8bac